building-an-api (#7)
Ada-Developers-Academy/core-unit-2: building-an-api/models.md

# Models

## Goal

Our goal for this lesson is:

1. Introduce the vocabulary "model," as it's commonly used in web development, and will be used in Flask
1. Briefly introduce the steps to setting up a model
1. Introduce the vocabulary "database migration" and the pattern of migrating the database

## Vocabulary and Synonyms

| Vocab              | Definition                                                                                                                                            | Synonyms                                   | How to Use in a Sentence                                                                                                                                  |
| ------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Model              | A representation of a single concept relevant to the application that contains the state and behavior                                                 | Data model, resource, entity, domain model | "Example models in an e-shopping web app might be `User`, `Product`, `Review`, and `Order`"                                                               |

Reviewer: "the state and behavior" feels a little disconnected

```suggestion
| Model              | A representation of a single concept relevant to the application. The model defines the state and behavior for that concept.                          | Data model, resource, entity, domain model | "Example models in an e-shopping web app might be `User`, `Product`, `Review`, and `Order`"                                                               |
```

| Database Migration | When working with relational databases, a representation of a change in a database's schema. Migrations must be created and then applied to databases | Schema migration, migration                | "We want the `Product` table to add a new integer column, `stock`. We need to generate a migration to represent that change, and then run the migrations" |

## Models

In software, a **model** is a representation of a single concept relevant to the application. This representation includes state and behavior.

For example, if we were building an app that represents a ride share service, we might have the models:

- `Driver`
- `Passenger`
- `Trip`

If "models" sounds similar to _resource_, _entity_, or _class_, you're onto something! All four of these terms can mean similar ideas, where the only notable differences are context.

## Models in Flask

We will define models in our Flask application code. We will use the package `SQLAlchemy` and follow their patterns to define and use models. Models in our Flask code will create a direct connection between the data modeled in our database, and the OOP Python code we can use in our back-end API.

> Therefore, when we set up each model, we will need to:
> 
> 1. Ensure our database and tables are aligned with our model
> 1. Ensure our Flask code has defined and can use the model

Reviewer: part of removing table stuff from the lesson. The next section provides a set of steps that we can expand to cover everything, so just cut this

```suggestion
```

> ## Models Relate to Databases

Reviewer: retitle since we won't make the same python/db split (since the db-specific section now would just be create a db)

```suggestion
## Models Are a Link Between a Database and Code
```

Our models are usually pieces of data that should be stored in a database. Their data should be persisted. To set up a model, we will need to:

1. Set up the database for the project

> 1. Set up the table for the model
> 1. Use our database migration tools

Reviewer: expand the steps to reflect the non-table focused version

```suggestion
1. Configure our app to connect to the database
1. Define our model in Python code
1. Ensure that Flask and SQLAlchemy are able to see our model code
1. Use tools to convert the model into instructions for creating or updating tables
1. Apply the instructions to our database to create or update the tables
```

### Creating the Database

Before we work with models in Flask, we should expect to create a database. We should continue to practice our great database skills!

> ### Creating the Tables
> 
> Similarly, before we work with different tables that hold our model data in Flask, we'll need to define database tables.

Reviewer 2: If my understanding from Ansel, that this isn't necessary or helpful, is true, we can remove this
```suggestion
```

> ### Connecting the DB and Flask

Reviewer: ```suggestion
### Connecting the Database and Flask
```

Flask needs a way to connect to our database. We'll do so by providing a path to it.

> ## Models Relate to Flask
> 
> In order for our Flask app to access the model, we will need to:
> 
> 1. Define the model class in Flask
> 1. Configure our Flask app to use the model

Reviewer: We don't need this separate chunk of steps anymore since we have the full layout above
```suggestion
```

> ## Create the Model File and Class

Reviewer: a few of the headers below here will need to be shifted in one since they should all be at the same level as the other steps
```suggestion
### Create the Model File and Class
```

We will create a class for each model. The class will define the state and behavior of our model.

### The `app/models` Directory

> SQLAlchemy and Flask patterns may encourage us to define each model as a class in its own file.

Reviewer 2: If there's a reason for the passivity in voice here, I'm open to it, but I'd tend to prefer a little stronger voice from the curriculum for reassurance that it does indeed know what its talking about
```suggestion
SQLAlchemy and Flask patterns encourage us to define each model as a class in its own file.
```

Reviewer: (I incorporated this suggestion into mine as well.)

> ### The `app/models` Directory
> 
> SQLAlchemy and Flask patterns may encourage us to define each model as a class in its own file.
> 
> This curriculum's recommended file structure recommends creating a `models` folder inside of the `app` folder. Note the location of a hypothetical model, `model`, in `app/models/model.py`

Reviewer: incorporates Jared's change
expands the description of the file layout to include the __init__.py needed for packaging

```suggestion
#### The `app/models` Directory

SQLAlchemy and Flask patterns encourage us to define each model as a class in its own file.

This curriculum's recommended file structure creates a `models` folder inside of the `app` folder. Note the location of a hypothetical model, `Model`, in `app/models/model.py`, as well as the typical `__init__.py` file to mark the folder as a package.
```

```
.

> ├── README.md
> ├── app
> │   ├── __init__.py
> │   ├── models
> │   │   └── model.py
> │   └── routes.py
> └── requirements.txt

Reviewer: reorder to reflect how VSCode would show things, and include the __init__.py in models
```suggestion
├── app
│   ├── models
│   │   ├── __init__.py
│   │   └── model.py
│   ├── __init__.py
│   └── routes.py
├── README.md
└── requirements.txt
```

```

> ### Configuring Models in Flask
> 
> Every time we define a new model, we'll need to make a small configuration update to our app.

Reviewer 2: This is among my weaker opinions, so take it or leave it, but I don't find this practical kind of information helpful to name until we're explaining what that configuration change is. As a student, I just don't know what to do with this information when I read it here.
```suggestion
```

> 

Reviewer: add additional sections for the generate migrations and apply migrations at this level of detail (I know it's brought up again below, but I think it works ok)

```suggestion

### Generate Instructions for Modifying Our Database

Flask and SQLAlchemy, our package connecting our models to the database, provide many useful model-related features. One such feature is that it can examine our models defined in Python, and produce a list of instructions for how to modify the database to be able to store the state of those models. This can include both creating new tables, and altering existing tables.

### Apply the Instructions to Modifying Our Database

Once the instructions have been generated, we use Flask and SQLAlchemy to apply those instructions to our database.

Generating and applying these instructions in separate steps allows for teams to work together. One member might add a new model, and generate instructions for a new table to be created. They will commit that model and the instructions to the team repo. The other members can retrieve those instructions and apply the same change to their local working databases.

```

> ### Using Defined Models in Flask
> 
> SQLAlchemy, our package that connects our models to the database, has powerful functionality. Instead of writing full lines of SQL, we will write Python that uses SQLAlchemy. SQLAlchemy will do most of the SQL for us!

Reviewer: bring this up a level to make a separate topic for using the models
some rewording since I used some of the existing verbiage above

```suggestion
## Using Defined Models in Flask

We don't need to write the code for storing, retrieving, or updating our models. SQLAlchemy lets us define our models as normal Python classes. Instead of writing full lines of SQL ourselves, we will write Python classes following patterns defined by SQLAlchemy, and SQLAlchemy will do most of the SQL heavy lifting for us!
```

SQLAlchemy has [already defined methods](https://flask-sqlalchemy.palletsprojects.com/en/2.x/queries/) for:

- Inserting new records into a table
- Querying records from a table
- Filtering records from a table
- Ordering records from a table
- Deleting records from a table

And much more!

> 

Reviewer: here's my suggestion for the ORM callout

```suggestion

### !callout-info

## SQLAlchemy Is an Object-Relational Mapper (ORM)

There are many ways to connect Python to a database. Some of them require us to write specific SQL queries ourselves. SQLAlchemy instead provides a connection from our Python classes (objects) to tables in a database (relations). In other words, it maps from object to relations. It's an Object-Relational Mapper!

<br />

In general, an Object-Relational Mapper is a tool that allows us to interact with tables without needing to write SQL directly.

### !end-callout

```

## Database Migrations

> On bigger and bigger projects, we should anticipate the effects of changing our database schemas. We can organize our database schema changes into _database migrations_.

Reviewer: provide motivation for why these migrations happen

```suggestion
On bigger and bigger projects, we should anticipate the need to change our database schemas. Over time we will need to add new models to our projects, or we might need to modify existing models. Changes to our models require changes to our database schema. To provide the ability to track our database schema changes, we can organize our changes into _database migrations_.
```

> A **datatbase migration** a representation of a change in a database's schema, usually as bit of code or a file of code. After migration files are created, these migration files can be applied to a database.

Reviewer: ```suggestion
A **database migration** is a representation of a change in a database's schema, usually as bit of code or a file of code. After migration files are created, these migration files can be applied to a database.
```

> Because the scale of our projects is increasing, we will also introduce a database migration tool.

Reviewer: I think the provided reason is spurious. Let's just say that if we are using migrations we need to use a migration tool.

```suggestion
Migrations are usually applied using a database migration tool.
```

> [Alembic](https://alembic.sqlalchemy.org/en/latest/) is the database migration tool that pairs nicely with SQLAlchemy. We will follow common database and Alembic practices to manage and maintain our migrations.

Reviewer: the original text would sound more grammatical as "Alembic is a database migration tool..." but I think declaring that it's the tool we will use and then cite the SQLAlchemy pairing flows a bit better. 

```suggestion
[Alembic](https://alembic.sqlalchemy.org/en/latest/) is the database migration tool that we will be using. It pairs nicely with SQLAlchemy. We will follow common database and Alembic practices to manage and maintain our migrations.
```

### One-time Setup: Initialize Migrations

> After the first database, table, and the configurations are set, we must initialize our migrations. We will only need to do this once.

Reviewer: excise table reference

```suggestion
After our database is created and we have configured our application to use it, we must initialize our migrations. We will only need to do this once.
```

> This command should create a new folder in our app, `migrations`.
> 
> ```
> .
> ├── README.md
> ├── app
> │   ├── __init__.py
> │   ├── models
> │   │   └── model.py
> │   └── routes.py
> ├── migrations
> │   ├── README
> │   ├── alembic.ini
> │   ├── env.py
> │   ├── script.py.mako
> │   └── versions
> └── requirements.txt
> ```
> 

Reviewer: This doesn't seem helpful here. We don't give the actual command, so why would we show the creation of the migration folder. Even when we _do_ give the command, I don't know that dwelling on the details of the contents provides much benefit.

```suggestion
```
